### PR TITLE
Add `JBuffer`, `JByteBuffer`

### PR DESCRIPTION
**Pull request by HosseinYousefi**
Speeds up working with buffers. 

@brianquinlan Here's a lean version of `JBuffer` and `JByteBuffer` APIs. Some changes have been made to the Java API to make it more Dart like. Such as making some methods return `void` due to the chaining operator in Dart (`..`) and using getters and setters instead of the traditional names. 

Some other methods have been left out which seemed not as useful. They can always be added in a later version.

TODOs:

- [x] Tests
- [x] Adding symbols to jni_symbols.yaml
- [x] Add a `fromList` factory to `JByteBuffer`

### CODE REVIEW

File: jni/lib/src/nio/jbyte_buffer.dart [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+        .callMethodWithArgs(
+            reference, _arrayId, JniCallType.objectType, []).object);
+  }
+
+  /// Returns this byte buffer as a [Uint8List].
+  ///
+  /// Throws [ArgumentError] if the buffer is not direct or the JVM does not
+  /// support the direct buffer operations.
+  Uint8List asUint8List() {
+    assert(isDirect, 'The buffer must be direct.');
```
Comment (brianquinlan) [inline jni/lib/src/nio/jbyte_buffer.dart:190]: if (!isDirect) {
  throw StateError('The buffer must be created with JByteBuffer.allocateDirect');
}

File: jni/lib/src/nio/jbyte_buffer.dart [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+    final address = Jni.env.GetDirectBufferAddress(reference);
+    if (address == nullptr) {
+      throw ArgumentError();
+    }
+    final capacity = Jni.env.GetDirectBufferCapacity(reference);
+    if (capacity == -1) {
+      throw ArgumentError();
```
Comment (brianquinlan) [inline jni/lib/src/nio/jbyte_buffer.dart:197]: StateError('The object is an unaligned view buffer and the processor architecture does not support unaligned access');

File: jni/lib/src/nio/jbyte_buffer.dart [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+      Jni.accessors.getMethodIDOf(_class.reference, r"array", r"()[B");
+
+  @override
+  JArray<jbyte> get array {
+    return const JArrayType(jbyteType()).fromRef(Jni.accessors
+        .callMethodWithArgs(
+            reference, _arrayId, JniCallType.objectType, []).object);
+  }
+
+  /// Returns this byte buffer as a [Uint8List].
```
[the remark this answers is not in the document]
Comment (HosseinYousefi) [reply jni/lib/src/nio/jbyte_buffer.dart:185]: > Eventually we could make it so that JByteBuffer implements List<Uint8>.

Yeah, this is nicer. I'll go with it.

[the remark this answers is not in the document]
Comment (dcharkes) [reply jni/lib/src/nio/jbyte_buffer.dart:185]: DBC

>  if the UInt8List is used after the java.nio.ByteBuffer is collected then you will be accessing deallocated memory

This is the typical reason why we have `Finalizable`s that wrap `Pointer`s.

To avoid extra copying, we don't want to eagerly copy from Java to Dart memory. So indeed, this class should indeed have the methods that use the `address` internally.

And it would make sense to have this class then implement `List<int>` indeed. 👍

[the remark this answers is not in the document]
Comment (HosseinYousefi) [reply jni/lib/src/nio/jbyte_buffer.dart:185]: From our discussion with @dcharkes:

`.asTypedList` has `finalizer` and `token` parameters and whenever the object is GCd, `finalizer` will be invoked for `token`. We can use this mechanism to create a safe `asUint8List` that removes a Java global reference when GCd.

`asUint8List` will also have an `releaseOriginal = false` flag that – when `true` – reuses the same global reference and sets the original object as released.

**Comment by HosseinYousefi** [comment]
I also included a `_asUint8ListUnsafe()` which does **not** attach any finalizer and can be used to add direct useful methods to `JByteBuffer`. We could consider having this:

Implement `List<int>` in `JByteBuffer`
1. With all methods using `_asUint8ListUnsafe()` under the hood and throwing error if the buffer is indirect.
2. With all methods checking `if (isDirect)` and handling the indirect case in a less efficient manner.

Right now I thought it's enough to provide `asUint8List()`, so e.g. `setRange` could be achieved like `buffer.asUint8List().setRange(...)` but this can make it shorter (and a tiny bit faster, if you want to keep the `JBuffer`) into `buffer.setRange(...)`.

File: jni/lib/src/nio/jbyte_buffer.dart [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+    return address.cast<Uint8>().asTypedList(
+          capacity,
+          token: token,
+          finalizer: Jni.env.ptr.ref.DeleteGlobalRef.cast(),
+        );
+  }
+}
+
+extension Uint8ListToJava on Uint8List {
+  /// Creates a [JByteBuffer] from the content of this list.
```
Comment (brianquinlan) [inline jni/lib/src/nio/jbyte_buffer.dart:258]: The [JByteBuffer] will be allocated using [JByteBuffer.allocateDirect].

File: jni/lib/src/nio/jbyte_buffer.dart [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+      Jni.accessors.callStaticMethodWithArgs(
+        _class.reference,
+        _wrapId,
+        JniCallType.objectType,
+        [array.reference, JValueInt(offset), JValueInt(length)],
+      ).object,
+    );
+  }
+
+  /// Creates a [JByteBuffer] from the content of [list].
```
Comment (brianquinlan) [inline jni/lib/src/nio/jbyte_buffer.dart:123]: The [JByteBuffer] will be created using [JByteBuffer.allocateDirect].

File: jni/lib/src/nio/jbyte_buffer.dart [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+    final address = _directBufferAddress();
+    final capacity = _directBufferCapacity();
+    return address.cast<Uint8>().asTypedList(capacity);
+  }
+
+  /// Returns this byte buffer as a [Uint8List].
+  ///
+  /// If [releaseOriginal] is `true`, this byte buffer will be released.
+  ///
+  /// Throws [StateError] if the buffer is not direct or the JVM does not
```
Comment (brianquinlan) [inline jni/lib/src/nio/jbyte_buffer.dart:238]: is not direct (see [JByteBuffer.allocateDirect]) or the ...

File: jni/pubspec.yaml [hunk 7]
```diff
@@ -4,7 +4,7 @@
 
 name: jni
 description: A library to access JNI from Dart and Flutter that acts as a support library for package:jnigen.
-version: 0.6.1
+version: 0.7.0-wip
```
Comment (brianquinlan) [inline jni/pubspec.yaml:7]: Could you do a release so that I can regenerate the files for `cronet_http`? We can also do that in a follow-up PR.

**Review by brianquinlan — CHANGES_REQUESTED** [review]

File: jni/lib/src/nio/jbyte_buffer.dart [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+      throw StateError(
+        'The buffer must be created with `JByteBuffer.allocateDirect`.',
+      );
+    }
+  }
+
+  Pointer<Void> _directBufferAddress() {
+    final address = Jni.env.GetDirectBufferAddress(reference);
+    if (address == nullptr) {
+      StateError(
```
[the remark this answers is not in the document]
Comment (HosseinYousefi) [reply jni/lib/src/nio/jbyte_buffer.dart:208]: oops!

**Comment by brianquinlan** [comment]
Nice! This is much better than the crap that I wrote originally ;-)

File: jni/lib/src/nio/jbyte_buffer.dart [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+  @override
+  int get hashCode => (JByteBufferType).hashCode;
+
+  @override
+  bool operator ==(Object other) {
+    return other.runtimeType == (JByteBufferType) && other is JByteBufferType;
+  }
+}
+
+class JByteBuffer extends JBuffer {
```
Comment (dcharkes) [inline jni/lib/src/nio/jbyte_buffer.dart:42]: This could benefit from a long doc comment:

* What it is (single sentence)
* Why to use it: fast memory copying between Java and Dart with `.setRange`.
* Example use snippets
   * How to use it with `asUint8List` with releaseOriginal true
   * How to use it with `asUint8List` with releaseOriginal false

File: jni/lib/src/nio/jbyte_buffer.dart [hunk 10]
```diff
@@ -0,0 +1,267 @@
...
+        JniCallType.objectType,
+        [array.reference, JValueInt(offset), JValueInt(length)],
+      ).object,
+    );
+  }
+
+  /// Creates a [JByteBuffer] from the content of [list].
+  ///
+  /// The [JByteBuffer] will be allocated using [JByteBuffer.allocateDirect].
+  factory JByteBuffer.fromList(List<int> list) {
```
Comment (dcharkes) [inline jni/lib/src/nio/jbyte_buffer.dart:126]: I think it's better to use Uint8List as a type here. I don't believe the `setAll` will be fast with `List<int>`.

Yes, it makes it slightly less convenient for users, but it will be fast. And then it's explicit for users that getting things from a generic `List<int>` is slow.

File: jni/lib/src/nio/jbyte_buffer.dart [hunk 11]
```diff
@@ -0,0 +1,267 @@
...
+  /// buffer's content.
+  JByteBuffer slice() {
+    return const JByteBufferType().fromRef(Jni.accessors.callMethodWithArgs(
+        reference, _sliceId, JniCallType.objectType, []).object);
+  }
+
+  static final _duplicateId = Jni.accessors.getMethodIDOf(
+      _class.reference, r"duplicate", r"()Ljava/nio/ByteBuffer;");
+
+  /// Creates a new byte buffer that shares this buffer's content.
```
Comment (dcharkes) [inline jni/lib/src/nio/jbyte_buffer.dart:145]: This could benefit from some explanation how this behaves as seen from a Dart programmer perspective.

It calls to Java to do `duplicate`.

Can we add the full documentation from Java instead of only the first line?

https://docs.oracle.com/javase/8/docs/api/java/nio/ByteBuffer.html#duplicate--

I'm not entirely sure how to combine the original docs from Java and the line that from Dart's perspective it's implemented by calling the Java method.

```dart
  /// Creates a new byte buffer that shares this buffer's content.
  ///
  /// Calls the Java method `duplicate` on this object.
  ///
  /// The content of the new buffer will be that of this buffer. Changes to this
  /// buffer's content will be visible in the new buffer; the new buffer itself,
  /// however, will be read-only and will not allow the shared content to be
  /// modified. The two buffers' position, limit, and mark values will be
  /// independent.
  ///
  /// The new buffer's capacity, limit, position, and mark values will be
  /// identical to those of this buffer.
  ///
  /// If this buffer is itself read-only then this method behaves in exactly the
  /// same way as the
  /// [duplicate](https://docs.oracle.com/javase/8/docs/api/java/nio/ByteBuffer.html#duplicate--)
  /// method.
  JByteBuffer duplicate() {
    return const JByteBufferType().fromRef(Jni.accessors.callMethodWithArgs(
        reference, _duplicateId, JniCallType.objectType, []).object);
  }
```

or

```dart
  /// Creates a new byte buffer that shares this buffer's content. (Calls the
  /// Java method `duplicate` on this object.)
  ///
  /// The content of the new buffer will be that of this buffer. Changes to this
  /// buffer's content will be visible in the new buffer; the new buffer itself,
  /// however, will be read-only and will not allow the shared content to be
  /// modified. The two buffers' position, limit, and mark values will be
  /// independent.
  ///
  /// The new buffer's capacity, limit, position, and mark values will be
  /// identical to those of this buffer.
  ///
  /// If this buffer is itself read-only then this method behaves in exactly the
  /// same way as the
  /// [duplicate](https://docs.oracle.com/javase/8/docs/api/java/nio/ByteBuffer.html#duplicate--)
  /// method.
  JByteBuffer duplicate() {
    return const JByteBufferType().fromRef(Jni.accessors.callMethodWithArgs(
        reference, _duplicateId, JniCallType.objectType, []).object);
  }
```

ditto for other methods from Java.

File: jni/lib/src/nio/jbyte_buffer.dart [hunk 12]
```diff
@@ -0,0 +1,262 @@
...
+  /// Returns this byte buffer as a [Uint8List].
+  ///
+  /// If [releaseOriginal] is `true`, this byte buffer will be released.
+  ///
+  /// Throws [StateError] if the buffer is not direct or the JVM does not
+  /// support the direct buffer operations or the object is an unaligned view
+  /// buffer and the processor does not support unaligned access.
+  Uint8List asUint8List({bool releaseOriginal = false}) {
+    _ensureIsDirect();
+    final address = _directBufferAddress();
```
[the remark this answers is not in the document]
Comment (dcharkes) [reply jni/lib/src/nio/jbyte_buffer.dart:246]: lgtm!

File: jni/test/jbyte_buffer_test.dart [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+      expect(buffer.remaining, 0);
+      expect(buffer.hasRemaining, false);
+    });
+  });
+
+  testRunner('isReadOnly and asReadOnlyBuffer', () {
+    using((arena) {
+      final buffer = testDataBuffer(arena);
+      expect(buffer.isReadOnly, false);
+      final readOnly = buffer.asReadOnlyBuffer();
```
Comment (dcharkes) [inline jni/test/jbyte_buffer_test.dart:140]: `final readOnly = buffer.asReadOnlyBuffer()..releasedBy(arena);` ?

File: jni/test/jbyte_buffer_test.dart [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+      expect(sliced.nextByte, 2);
+    });
+  });
+
+  testRunner('duplicate', () {
+    using((arena) {
+      final buffer = testDataBuffer(arena);
+      buffer.position = 1;
+      buffer.limit = 2;
+      final duplicate = buffer.duplicate();
```
Comment (dcharkes) [inline jni/test/jbyte_buffer_test.dart:181]: `..releasedBy(arena)`?

File: jnigen/lib/src/tools/android_sdk_tools.dart [hunk 15]
```diff
@@ -202,6 +202,7 @@ task $_gradleGetSourcesTaskName(type: Copy) {
       File(buildGradleOld).deleteSync();
     }
     if (procRes.exitCode != 0) {
+      print(procRes.stdout);
```
Comment (dcharkes) [inline jnigen/lib/src/tools/android_sdk_tools.dart:205]: left over debug print?

File: jnigen/lib/src/bindings/dart_generator.dart [hunk 16]
```diff
@@ -589,7 +589,7 @@ class _$implClassName$typeParamsDef implements $implClassName$typeParamsCall {
         ? '($typeClassName).hashCode'
         : 'Object.hash($typeClassName, $hashCodeTypeClasses)';
     s.write('''
-class $typeClassName$typeParamsDef extends $_jType<$name$typeParamsCall> {
+final class $typeClassName$typeParamsDef extends $_jType<$name$typeParamsCall> {
```
Comment (dcharkes) [inline jnigen/lib/src/bindings/dart_generator.dart:592]: Nit: could have been a separate PR to ease reviewing. (I have already reviewed now, so keep it in here!)

File: jnigen/CHANGELOG.md [hunk 17]
```diff
@@ -1,5 +1,7 @@
-## 0.6.1-wip
-* Add `ignore_for_file: lines_longer_than_80_chars` to the generated file preamble.
+## 0.7.0
+* **Breaking Change** ([#387](https://github.com/dart-lang/jnigen/issues/387)): Added `JBuffer` and `JByteBuffer` classes as default classes for `java.nio.Buffer` and `java.nio.ByteBuffer` respectively.
```
Comment (dcharkes) [inline jnigen/CHANGELOG.md:2]: nit: line wrap at 80?

File: jni/lib/src/util/jlist.dart [hunk 18]
```diff
@@ -196,7 +196,7 @@ class JList<$E extends JObject> extends JObject with ListMixin<$E> {
           reference,
           _insertAllId,
           JniCallType.booleanType,
-          [JValueInt(index), (iterable as JObject).reference]);
+          [JValueInt(index), (iterable as JObject).reference]).boolean;
```
Comment (dcharkes) [inline jni/lib/src/util/jlist.dart:199]: nit: prefer trailing comma for multi-line arguments.

**Review by dcharkes — APPROVED** [review]
lgtm!

Please add also a speedup factor to the PR description, our motivation! 💨

File: jni/lib/src/nio/jbyte_buffer.dart (hunk 11 above)
Comment (HosseinYousefi) [reply jni/lib/src/nio/jbyte_buffer.dart:145]: I'll address this in another PR. Other classes like also don't have any documentation. 

https://github.com/dart-lang/jnigen/issues/391

File: jni/lib/src/nio/jbyte_buffer.dart (hunk 10 above)
Comment (HosseinYousefi) [reply jni/lib/src/nio/jbyte_buffer.dart:126]: Fair enough.

File: jni/lib/src/nio/jbyte_buffer.dart (hunk 11 above)
Comment (HosseinYousefi) [reply jni/lib/src/nio/jbyte_buffer.dart:145]: Btw the docs you have here is for `asReadOnlyBuffer`.

**Comment by HosseinYousefi** [comment]
@dcharkes Please review the added docs.

File: jni/lib/src/nio/jbyte_buffer.dart [hunk 19]
```diff
@@ -0,0 +1,302 @@
...
+
+/// A byte [JBuffer].
+///
+/// The bindings for `java.nio.ByteBuffer`.
+///
+/// This enables fast memory copying between Java and Dart when directly
+/// allocated (See [JByteBuffer.allocateDirect]).
+///
+/// To create a [JByteBuffer] from the content of a [Uint8List],
+/// use [JByteBuffer.fromList]. This uses direct allocation and will greatly
```
Comment (dcharkes) [inline jni/lib/src/nio/jbyte_buffer.dart:50]: Maybe rephrase as "This uses direct allocation to enable fast copying".

Speed up is a comparison, but a comparison to what?

File: jni/lib/src/nio/jbyte_buffer.dart [hunk 20]
```diff
@@ -0,0 +1,302 @@
...
+/// speed up the copying.
+///
+/// [asUint8List] provides a direct access to the underlying [Uint8List] that
+/// this buffer uses. This means any changes to it will change the content of
+/// the buffer and vice versa. You can use this to access to [Uint8List] methods
```
Comment (dcharkes) [inline jni/lib/src/nio/jbyte_buffer.dart:55]: Avoid "you". This can be used to ...

File: jni/lib/src/nio/jbyte_buffer.dart [hunk 21]
```diff
@@ -0,0 +1,302 @@
...
+/// such as [Uint8List.setRange].
+///
+/// It is safe to use [asUint8List]. As long as the resulting list is
```
Comment (dcharkes) [inline jni/lib/src/nio/jbyte_buffer.dart:58]: The first sentence saying "safe" makes me wonder safe in what sense before I read on. Maybe rephrase to: "Both the original buffer and the Uint8List keep the underlying Java buffer alive. Once all the ..." (I think the right terms to use are "alive" and "release". Maybe don't use "safe" but spell it out that we will not try to read the Java memory from Dart after it has been "released".)

File: jni/lib/src/nio/jbyte_buffer.dart [hunk 22]
```diff
@@ -0,0 +1,302 @@
...
+/// accessible, the associated Java buffer will also be kept alive. Once all the
+/// instances of the original buffer and the lists produced from [asUint8List]
+/// are inaccessible both in Java and Dart, Java will correctly garbage collects
+/// the buffer and frees its underlying memory.
+///
+/// You can choose to [release] the original buffer when calling [asUint8List]
```
Comment (dcharkes) [inline jni/lib/src/nio/jbyte_buffer.dart:64]: another you

File: jni/lib/src/nio/jbyte_buffer.dart [hunk 23]
```diff
@@ -0,0 +1,302 @@
...
+///
+/// To create a [JByteBuffer] from the content of a [Uint8List],
+/// use [JByteBuffer.fromList]. This uses direct allocation and will greatly
+/// speed up the copying.
+///
+/// [asUint8List] provides a direct access to the underlying [Uint8List] that
+/// this buffer uses. This means any changes to it will change the content of
+/// the buffer and vice versa. You can use this to access to [Uint8List] methods
+/// such as [Uint8List.setRange].
+///
```
Comment (dcharkes) [inline jni/lib/src/nio/jbyte_buffer.dart:57]: I'd move the simplest version of the example here.

Please make the example also memory safe with using the arena or releasing manually. Users are going to copy and paste it.

File: jni/lib/src/nio/jbyte_buffer.dart [hunk 24]
```diff
@@ -0,0 +1,302 @@
...
+/// It is safe to use [asUint8List]. As long as the resulting list is
+/// accessible, the associated Java buffer will also be kept alive. Once all the
+/// instances of the original buffer and the lists produced from [asUint8List]
+/// are inaccessible both in Java and Dart, Java will correctly garbage collects
+/// the buffer and frees its underlying memory.
+///
```
Comment (dcharkes) [inline jni/lib/src/nio/jbyte_buffer.dart:63]: Then a second example here, where we release the buffer early, and access the Uint8List after.

File: jni/lib/src/nio/jbyte_buffer.dart [hunk 25]
```diff
@@ -0,0 +1,302 @@
...
+/// You can choose to [release] the original buffer when calling [asUint8List]
+/// by setting the `releaseOriginal` parameter to `true`.
+///
```
Comment (dcharkes) [inline jni/lib/src/nio/jbyte_buffer.dart:66]: Then third example here where we use `releaseOriginal: true`.

**Review by dcharkes — APPROVED** [review]
